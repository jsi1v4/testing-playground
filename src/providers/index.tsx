import React, { createContext, useContext } from "react";

import { Service } from "./mock-service";
import { IService } from "./types";

export const Context = createContext<IService>(null as never);

export const useTestContext = () => useContext(Context);

export function TestProvider({ children }) {
  return <Context.Provider value={new Service()}>{children}</Context.Provider>;
}
