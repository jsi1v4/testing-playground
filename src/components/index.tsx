import React from "react";

import { Data } from "../providers/types";

export interface TableProps {
  data?: Data[];
}

export const Table: React.FC<TableProps> = ({ data }) => {
  return (
    <table data-testid="id-table">
      <thead>
        <tr>
          <th>Letter</th>
          <th>Number</th>
        </tr>
      </thead>
      <tbody>
        {data?.map((row, key) => (
          <tr key={key}>
            <td data-testid={`id-table-letter-${key}`}>{row.letter}</td>
            <td data-testid={`id-table-number-${key}`}>{row.number}</td>
          </tr>
        ))}
      </tbody>
    </table>
  );
};
