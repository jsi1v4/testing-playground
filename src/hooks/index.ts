import { useCallback, useEffect, useState } from "react";

import { useTestContext } from "../providers";
import { Data } from "../providers/types";

export function useTestHook() {
  const service = useTestContext();

  const [data, setData] = useState<Data[]>();

  const fetch = useCallback(() => {
    return service.getInfos().then(setData).catch(console.error);
  }, [service]);

  useEffect(() => {
    fetch();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return {
    data,
  };
}
