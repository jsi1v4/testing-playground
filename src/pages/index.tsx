import React from "react";
import { NextPage } from "next";

import { useTestHook } from "../hooks";
import { Table } from "../components";

export const Home: NextPage = () => {
  const { data } = useTestHook();

  return <Table data={data} />;
};

export default Home;
