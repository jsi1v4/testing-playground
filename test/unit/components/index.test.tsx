import React from "react";
import { render } from "@testing-library/react";

import { Table } from "../../../src/components";

const MOCK = [
  {
    letter: "A",
    number: 1,
  },
];

describe("unit-test Table component", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(<Table data={MOCK} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render component", async () => {
    const { getByTestId } = render(<Table data={MOCK} />);
    expect(getByTestId("id-table").children).toHaveLength(2);
    expect(getByTestId("id-table-letter-0").innerHTML).toBe("A");
    expect(getByTestId("id-table-number-0").innerHTML).toBe("1");
  });
});
