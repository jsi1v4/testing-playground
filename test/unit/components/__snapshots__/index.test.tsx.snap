// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit-test Table component matches snapshot 1`] = `
<DocumentFragment>
  <table
    data-testid="id-table"
  >
    <thead>
      <tr>
        <th>
          Letter
        </th>
        <th>
          Number
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          data-testid="id-table-letter-0"
        >
          A
        </td>
        <td
          data-testid="id-table-number-0"
        >
          1
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
